### PR TITLE
Add SerializedNameConvention annotation and metadata

### LINKED ISSUE #29: Add serialized naming scheme annotation
Currently, if you want to use another naming scheme in config files than in the java source code, you need to specify and separately update an `@SerialzedName`  annotation. In order to fix this, I suggest a naming scheme annotation which allows to specify a serialized naming scheme for the given config value, section, or config.

It should work similar to #4 's `NameConvention` annotation, however `SerializedNameConvention` is a better match imo 
Ideally, it accepts a NamingScheme enum `value`, or a `custom` class, which implements the `NamingScheme` interface

### PR DESCRIPTION
**Pull request by anonymous123-code**
Metadata types can be made inhertitable by using true in the create methods (There are overloads defaulting to false)
When the config builder is finally built, all inheritable metadata is propagated and applied to its children.
If children specify their own metadata, that takes precedence.

Additionally makes it so that ConfigFieldAnnotationProcessors can also process the config class's annotations.

Specific to SerializedNameConvention
- it is inherited
- Adds a serialized naming scheme annotation, allowing for automatic case conversion for serialization
- The `NamingScheme` interface and the `NamingSchemes` enum, inspired by Configurate, but more flexible to prepare for #30 
- Both `NameConvention.custom()` and `NameConvention.value()` can be set at the same time, but `NameConvention.custom()` takes priority.
- Sections inherit the main config class's `NameConvention` by default.
- Sections can have different `NameConventions` for their nested values. A `NameConvention` annotation to the section field, with the annotation on the field taking priority.
- The default naming scheme is `PASSTHROUGH`, for backwards compatibility.

Based on #4 
Fixes #29

### CODE REVIEW

**Comment by ix0rai** [comment]
please run `applyLicenses`!

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append(word);
+					}
+				}
+			}
+			return builder.toString();
+		}
+
+	},
+	/**
+	 * Formats the string as title case, all words are capitalized. {@code "funnyStringAsAnExample"} becomes {@code "Funny String As An Example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:158]:
```suggestion
	 * Formats the string as {@code Title Case}, all words are capitalized. {@code "funnyStringAsAnExample"} becomes {@code "Funny String As An Example"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append(" ");
+					}
+					builder.append(word);
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as space separated lower case with the initial letter being uppercase. {@code "funnyStringAsAnExample"} becomes {@code "Funny string as an example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:134]:
```suggestion
	 * Formats the string as {@code Space separated lower case} with the initial letter being uppercase. {@code "funnyStringAsAnExample"} becomes {@code "Funny string as an example"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append("_");
+					}
+					builder.append(word);
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as space separated lower case. {@code "funnyStringAsAnExample"} becomes {@code "funny string as an example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:113]:
```suggestion
	 * Formats the string as {@code space separated lower case}. {@code "funnyStringAsAnExample"} becomes {@code "funny string as an example"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append("-");
+					}
+					builder.append(word);
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as snake_case. {@code "funny string as an example"} becomes {@code "funny_string_as_an_example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:92]:
```suggestion
	 * Formats the string as {@code snake_case}. {@code "funny string as an example"} becomes {@code "funny_string_as_an_example"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append(Character.toUpperCase(word.codePointAt(0)));
+						builder.append(word.substring(1));
+					}
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as kebab-case. {@code "funny string as an example"} becomes {@code "funny-string-as-an-example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:71]:
```suggestion
	 * Formats the string as {@code kebab-case}. {@code "funny string as an example"} becomes {@code "funny-string-as-an-example"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+				if (!word.isEmpty()) {
+					builder.append(Character.toUpperCase(word.codePointAt(0)));
+					builder.append(word.substring(1));
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as lowerCamelCase. {@code "funny string as an example"} becomes {@code "funnyStringAsAnExample"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:49]:
```suggestion
	 * Formats the string as {@code lowerCamelCase}. {@code "funny string as an example"} becomes {@code "funnyStringAsAnExample"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	 * Does not format the string at all. {@code "funnyStringAsAnExample"} becomes {@code "funnyStringAsAnExample"}
+	 */
+	PASSTHROUGH {
+		@Override
+		public String coerce(String input) {
+			return input;
+		}
+	},
+	/**
+	 * Formats the string as UpperCamelCase. {@code "funny string as an example"} becomes {@code "FunnyStringAsAnExample"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:33]:
```suggestion
	 * Formats the string as {@code UpperCamelCase}. {@code "funny string as an example"} becomes {@code "FunnyStringAsAnExample"}
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingScheme.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.config.api.metadata;
+
+/**
+ * A naming scheme to indicate how a string should be formatted.
+ * <p>
+ * Common formats are exposed in {@link NamingSchemes}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingScheme.java:22]:
```suggestion
 * Common formats are exposed in {@link NamingSchemes}.
```

File: src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+import org.quiltmc.config.api.metadata.NamingSchemes;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Optional;
+
+/**
+ * Used to tell the serializer how properties should be formatted when saving to disk. Can be applied to configs, sections and properties. {@link SerializedName} should always take priority.
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java:31]:
```suggestion
 * Used to tell the serializer how properties should be formatted when saving to disk. Can be applied to configs, sections and properties. {@link SerializedName} will always take priority.
```

Comment (ix0rai) [reply src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java:31]: confidence!

**Review by ix0rai — COMMENTED** [review]
javadoc nitpicks! all code looks good, though I'd like to see tests for this before merging

File: src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java (hunk 9 above)
Comment (ix0rai) [reply src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java:31]: must is much better, yes!

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java (hunk 3 above)
Comment (ix0rai) [reply src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:113]: I'd still apply these for consistency's sake

File: src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java (hunk 9 above)
Comment (ix0rai) [reply src/main/java/org/quiltmc/config/api/annotations/SerializedNameConvention.java:31]:
```suggestion
 * Used to tell the serializer how properties should be formatted when saving to disk. Can be applied to configs, sections and properties. {@link SerializedName} must always take priority.
```

File: src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+						builder.append(" ");
+					}
+					builder.append(word);
+				}
+			}
+			return builder.toString();
+		}
+	},
+	/**
+	 * Formats the string as space separated lower case with the first word being capitalized. {@code "funnyStringAsAnExample"} becomes {@code "Funny string as an example"}
```
Comment (ix0rai) [inline src/main/java/org/quiltmc/config/api/metadata/NamingSchemes.java:134]:
```suggestion
	 * Formats the string as {@code Space separated lower case} with the first word being capitalized. {@code "funnyStringAsAnExample"} becomes {@code "Funny string as an example"}
```

**Review by ix0rai — COMMENTED** [review]
popping in some suggestions
I do think that metadata added using processors should have the same behaviour as metadata added via annotations!

**Comment by ix0rai** [comment]
@anonymous123-code any plans to finish this?

**Review by ix0rai — APPROVED** [review]
I meticulously pored over this and yet could not find any issues. great PR!

**Review by TheGlitch76 — APPROVED** [review]
One small thing

**Pull request merged**